NEUT-1765 Add Sponsored Banner SwiftUI Component

darioroa opened the pull request:
<!--
Thanks for contributing to Backpack :pray:

Please include a description of the changes you are introducing and some screenshots if appropriate.
-->

This PR adds the Sponsored Banner Backpack Component

![image](https://github.com/Skyscanner/backpack-ios/assets/26660989/09e86e18-30c7-4506-a758-72899dfb5188)


+ [x] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-ios/blob/main/CONTRIBUTING.md)

Remember to include the following changes:
+ [x] `README.md`
+ [x] Tests
+ [x] [Screenshotting code](https://github.com/Skyscanner/backpack-ios/blob/main/Example/Backpack%20Screenshot/Screenshots.swift)
+ [ ] Adding a component? Remember to expose it in the [main `Backpack.h` header file](https://github.com/Skyscanner/backpack-ios/tree/main/Backpack/Backpack.h)

_If you are curious about how we review, please read through the [code review guidelines](https://github.com/Skyscanner/backpack/blob/main/CODE_REVIEW_GUIDELINES.md)_

AxelCB commented on Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:
```diff
@@ -0,0 +1,171 @@
+    
+    private var iconView: some View {
+        ZStack {
+            Text("")
```
Why do we need an empty text?

darioroa replied:
This is a workaround due to the VoiceOver issues. Investigating.

darioroa replied:
It's nicely done now 👍

AxelCB commented on Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:
```diff
@@ -0,0 +1,171 @@
+        .accessibilityAddTraits(.isButton)
+        .accessibilityValue(callToAction?.accessibilityLabel ?? "")
+        .frame(height: 22)
+        .frame(height: 22)
```
Double frame modifier? 🤔

darioroa replied:
Sorted 👍

AxelCB commented on Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:
```diff
@@ -0,0 +1,171 @@
+                .foregroundColor(variant.color)
+        }
+        .accessibilityAddTraits(.isButton)
+        .accessibilityValue(callToAction?.accessibilityLabel ?? "")
```
Does accessibility value with an empty string make sense? Can't we use `ConditionalModifier` here?

darioroa replied:
Did a similar thing, check latest PR 😄

AxelCB commented on Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:
```diff
@@ -0,0 +1,171 @@
+    
+    private var topView: some View {
+        HStack(spacing: .md) {
+            HStack {
```
Why do we need 2 Hstack levels here?

darioroa replied:
Good catch, initially it was to have separation of the info label from the rest, but perhaps it's not needed anymore 🤔

darioroa replied:
Done, check 98da5d5e583a83f8c6ac8ddbf181a0c07f300667

AxelCB commented on Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:
```diff
@@ -0,0 +1,171 @@
+    let variant: Variant
+    let backgroundColor: Color
+    
+    @State var showTerms: Bool
```
Personally I find this name feeling a bit incomplete, I'd be okay with a longer more descriptive name `showTermsAndConditions` but happy to hear what others think on short vs long names here 😄

AxelCB commented on Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:
```diff
@@ -0,0 +1,171 @@
+            subheadline: "Subheading",
+            callToAction: .init(
+                text: "Sponsored",
+                accessibilityLabel: "More Information", showIcon: false),
```
should showIcon be on a separate line?

darioroa replied:
Yes!

AxelCB reviewed (commented):
Besides the inline comments, got a couple of questions:

1. Why are you adding duplicated snapshots under BackpackTests/FailureDiffs?
2. Should we include tests/screenshots for a sponsored banner without text/title?
3. For dark mode test & examples with bodyText/ terms and conditions, I'd consider if we should wrap the component around with a view with another colour, or even if the chosen one is correct since there's no contrast with the default canvas colour

darioroa commented:
> Besides the inline comments, got a couple of questions:
> 
> 1. Why are you adding duplicated snapshots under BackpackTests/FailureDiffs?

Was a mistake, corrected by @gert-janvercauteren (thank you!)

> 2. Should we include tests/screenshots for a sponsored banner without text/title?

Thought about it, in fact it's the "default" initialiser, but it does not make sense implementation-wise. It's just for the case where there is, for instance, only a logo on the left and a "sponsored" text on the right (for Inline plus, for instance).

> 3. For dark mode test & examples with bodyText/ terms and conditions, I'd consider if we should wrap the component around with a view with another colour, or even if the chosen one is correct since there's no contrast with the default canvas colour

Corrected, you can take a look at the updated custom token for better contrast 👍

gert-janvercauteren commented on Backpack-SwiftUI/Card/Classes/BPKCardWrapper.swift:
```diff
@@ -36,6 +36,16 @@ public struct BPKCardWrapper<HeaderContent: View, WrappedContent: CardWrappable>
         self.card = card()
     }
     
+    public init(
+        backgroundColor: UIColor,
+        @ViewBuilder header: () -> HeaderContent,
+        @ViewBuilder card: () -> WrappedContent
+    ) {
+        self.backgroundColor = BPKColor.init(value: backgroundColor)
+        self.header = header()
+        self.card = card()
+    }
```
Is this change needed in this PR?

darioroa replied:
Definitely not, made it into it somehow when merging stuff 🤔 . Checking 👍

gert-janvercauteren commented on Backpack-SwiftUI/Card/Classes/BPKCardWrapper.swift:
```diff
@@ -102,6 +112,22 @@ struct BPKCardWrapper_Previews: PreviewProvider {
                 }
             }
+            
+            BPKCardWrapper(backgroundColor: UIColor.magenta) {
+                HStack(spacing: 5) {
+                    BPKIconView(.flag)
+                    BPKText("Custom Color", style: .caption)
+                        .foregroundColor(.white)
+                    Spacer()
+                }
+                .foregroundColor(.white)
+                .padding(.md)
+            } card: {
+                BPKCard(elevation: .none) {
+                    Text("Card Content")
+                }
+            }
```
This probably needs to be rolled back

gert-janvercauteren commented on Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:
```diff
@@ -0,0 +1,201 @@
+    let backgroundColor: Color
+    let customAccessibilityLabel: String?
+    
+    @State var showTerms: Bool
```
This can probably be "isExpanded" as that is what it does

frugoman replied:
This! ☝️

gert-janvercauteren commented on Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:
```diff
@@ -0,0 +1,201 @@
+import SwiftUI
+
+public struct BPKSponsoredBanner: View {
+    let logo: Image?
+    let title: String?
+    let subheadline: String?
+    let callToAction: CallToAction?
+    let bodyText: String?
+    let variant: Variant
+    let backgroundColor: Color
+    let customAccessibilityLabel: String?
+    
+    @State var showTerms: Bool
```
These probably all need to be private

gert-janvercauteren commented on Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:
```diff
@@ -0,0 +1,201 @@
+    
+    public struct CallToAction {
+        public let text: String
+        public let accessibilityLabel: String
```
Please rename this to a11y hint,
do we think we need this in general if we can 'expand' the banner?

darioroa replied:
🤔 Good point. This is more of a product/design level req, but if there is already an `isButton` trait, it's clear it is expandable even without this.

gert-janvercauteren commented on Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:
```diff
@@ -0,0 +1,201 @@
+            subheadline: "Subheading",
+            callToAction: .init(
+                text: "Sponsored",
+                accessibilityLabel: "More Information",
```
"Double tap for more information"

gert-janvercauteren reviewed (commented):
Left some remarks on the accessibility / general build.

frugoman commented on Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:
```diff
@@ -0,0 +1,201 @@
+        }
+    }
+    
+    private var bodyTextView: some View {
+        Group {
```
can you make this a `@ViewBuilder` and remove the `Group`? that way if the text does not need to be shown nothing would be added

frugoman commented on Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:
```diff
@@ -0,0 +1,201 @@
+    
+    private var bodyTextView: some View {
+        Group {
+            if hasBody && showTerms == true {
+                BPKText(bodyText ?? "", style: .caption)
```
I think you probably want to also if-let the body text as well, to not accidentally show an empty BPKText

frugoman commented on Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:
```diff
@@ -0,0 +1,201 @@
+                BPKText(callToAction.text, style: .caption)
+                    .foregroundColor(variant.color)
+            }
+            if callToAction?.showIcon == true {
```
should this line be inside the if let thats right above?

darioroa replied:
Good catch!

frugoman commented on Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:
```diff
@@ -0,0 +1,201 @@
+    }
+}
+
+extension BPKSponsoredBanner {
```
could you move this extension to a new file? nothing bad to have it here really, just to make the files smaller and easier to read later

frugoman commented on Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:
```diff
@@ -0,0 +1,201 @@
+        case onLight
+        case onDark
+        
+        var color: Backpack_SwiftUI.BPKColor {
+            (self == .onDark) ? BPKColor.textOnDarkColor : BPKColor.textOnLightColor
```
I think you can swftify this props (if you still need them for this to compile then it's fine)
```swift
var color: BPKColor {
  self == .onDark ? .textOnDarkColor : .textOnLIghtColor
}

darioroa commented on Backpack-SwiftUI/SponsoredBanner/Classes/BPKSponsoredBanner.swift:
```diff
@@ -0,0 +1,201 @@
+    
+    @State var showTerms: Bool
+    
+    public init(
+        logo: Image? = nil,
+        title: String? = nil,
+        subheadline: String? = nil,
+        callToAction: CallToAction? = nil,
+        bodyText: String? = nil,
+        variant: Variant,
+        backgroundColor: Color,
+        showTerms: Bool = false,
+        customAccessibilityLabel: String? = nil
```
Completely agree, let's check with Design about which ones are possible states so we can properly prevent them.

frugoman reviewed (approved):
Looks pretty good!! thanks for contributing it!!
